experimental/incremental/task.go

```diff
@@ -339,177 +335,179 @@ type Result[T any] struct {
 	Changed bool
 }
 
-// result is a Result[any] with a completion channel appended to it.
-type result struct {
-	Result[any]
-
-	// This is the sequence ID of the Run call that caused this result to be
-	// computed. If it is equal to the ID of the current Run, it was computed
-	// during the current call. Otherwise, it is cached from a previous Run.
-	//
-	// Proof of correctness. As long as any Runs are ongoing, it is not possible
-	// for queries to be evicted, so once a query is calculated, its runID is
-	// fixed. Suppose two Runs race the same query. One of them will win as the
-	// leader, and the other will wait until it's done. The leader will mark it
-	// with its run ID, so the leader sees Changed and the loser sees !Changed.
-	// Any other queries from the same or other Runs racing this query will see
-	// the same result.
-	//
-	// Note that runID itself does not require synchronization, because loads of
-	// runID are synchronized-after the done channel being closed.
-	runID uint64
-	done  chan struct{}
-}
-
-// path is a linked list node for tracking cycles in query dependencies.
-type path struct {
-	Query *AnyQuery
-	Prev  *path
-}
-
-// Walk returns an iterator over the linked list.
-func (p *path) Walk() iter.Seq[*path] {
-	return func(yield func(*path) bool) {
-		for node := p; node.Query != nil; node = node.Prev {
-			if !yield(node) {
-				return
-			}
-		}
-	}
-}
-
 // start executes a query in the context of some task and records the result by
 // calling done.
 //
 // If sync is false, the computation will occur asynchronously. Returns whether
 // the computation is in fact executing asynchronously as a result.
-func (t *task) start(caller *Task, q *AnyQuery, sync bool, done func(*result)) (async bool) {
-	// Common case for cached values; no need to spawn a separate goroutine.
-	r := t.result.Load()
-	if r != nil && closed(r.done) {
-		caller.log("cache hit", "%T/%v", q.Underlying(), q.Underlying())
-		done(r)
+func (t *Task) start(query *AnyQuery, sync bool, done func(*task)) (async bool) {
+	t.exec.lock.Lock()
+	if t.exec.tasks == nil {
+		t.exec.tasks = make(map[any]*task)
+	}
```

Comment:
what happens if two gs race to set this value? technically map pointers are tear-free in all architectures we care about, but this remains... janky